### ISSUE TITLE
Constant-time padding checks do not seem usefully constant-time

### ISSUE THREAD

**Issue by davidben** [pyca/cryptography#5871]
There are a bunch of constant-time padding check implementations in `src/rust` (and previously in C), but they're only used by `_byte_unpadding_check` which, immediately after checking the padding, branches on the result and sizes buffers based on the final length.

I was not able to find any reference in the code for what attack this stops. From what I can tell, it doesn't seem to actually provide any useful side channel mitigations. If the ciphertext was already authenticated before being decrypting and unpadding, then it may as well have been done in variable-time. The only information there was to leak is the length, and `_byte_unpadding_check` leaks the length.

If the ciphertext wasn't authenticated, it's very unlikely that the constant-time padding check helps. Where protocols have gotten the encryption and MAC ordering wrong, the fix has been more complex and requires integrating the padding check with a MAC check. (See, for example, the Lucky 13 fix.) The APIs exposed here are not sufficient to implement this and will leak if used as part of such an integrated check. Where such things are needed, they're more the purview of the underlying cryptographic library, which has more room to set up the right abstractions.

What attack or information leak was this meant to avoid?

**Comment by alex** [comment]
Gosh it really would be nice if I remembered the precise answer to this question, they've been this way for as long as we've had unpadding.

I _think_ the answer is "Lucky 13 except we didn't think the abstraction all the way through". Once upon a time we were interested in the idea of supporting TLS libraries being implemented in Python, in which case we would have needed this code with a better API.

Of course, nowadays the CBC TLS ciphersiutes are basically dead (ding dong teh witch is dead!).

In conclusion... maybe there's not a good reason. It was fun code to write though.

@reaperhulk do you remember better than me?

**Comment by davidben** [comment]
If it's specifically for Lucky13, this code wouldn't help. You'd need to not only check padding but also return the amount of padding. That then can't be returned through Python because transiting integers through Python leaks. You'd need the C/Rust API, in one go, compute padding, call into OpenSSL to compute the expected MAC (over a secret prefix), and extract the received MAC (from a secret location).

All that is enough of a mess that it's better modeled as a funny authenticated encryption API that the underlying crypto library exposes. (We do that in BoringSSL. Our CBC mode TLS ciphers are packaged up as funny AEADs.)

**Comment by reaperhulk** [comment]
Yeah, as I recall this was originally put in place to try to mitigate unauthenticated classical padding oracles and nothing else. Not from even incorporating them into larger protocols, just the most basic "implement a crypto 101 attack" concept. I'm a bit reluctant to say this is valueless, but I'm open to being convinced otherwise.

**Comment by th3b0x** [comment]
@davidben I hate to be a lazy slug here, but could you cite a few examples of the padding checks you're referencing?

I have a few on-hand reference materials that might be able to provide a specific notable padding oracle attack (or block cipher exploits related to improper utilization/implementation).  However, a lot of this stuff is contained in physical notebooks and the textbooks are "lock, key, and secure room" tier, so it helps when someone can designate some example "targets" in advance before I start pealing through handwritten notes in wide-ruled spiral-bound notebooks.

I can't articulate the specific rationale here, but my mental notion is that the behavior described is standard practice for sane cryptography, particularly with relation to block ciphers, stream ciphers, and more elaborate schemes utilizing ECC or embedded message digests.

From a runtime performance or architectural analysis perspective, parallel processing of a fixed-size buffer/block primitive is a SIMD (single instruction multiple data) situation, so implementing this in constant time is always going to be useful (particularly in circumstances when the underlying procedure is frequently utilized).

I am not going to *speculate* as to the intended purposes, but I cannot - for a second - imagine that there is anything "_un_usefully constant-time" about these functions.

I feel like I'm preaching to the choir, but I guess my question - at this point - has to do with why you're intrigued by the seemingly-unnecessary presence of this code.  Is it a concern about bit rot, bit bloat, or is the mystery just tickling your Reverse Engineering sense?  Perhaps I am missing a key detail here because I lack detailed insight into the `Rust` system's internals.  I would immediately agree with your concern if this _were_ _not_ specifically about `src/rust` .  The Rust data internal mechanisms are significantly more sophisticated than what is offered by vanilla C and C++, or the libraries commonly utilized for memory/data handling in those languages.

**Comment by alex** [comment]
Going to close this based on @reaperhulk's last comment. In conclusion everyone should just use AEADs and then we won't have to care.
